src/Client/PacketReceiver.h

```diff
@@ -0,0 +1,225 @@
+            {
+                if (errno == EAGAIN)
+                    break;
+
+                if (errno != EINTR)
+                    throwFromErrno("Cannot write to pipe", ErrorCodes::CANNOT_READ_FROM_SOCKET);
+            }
+        }
+
+        /// Receiving packet was finished.
+        return std::move(packet);
+    }
+
+    void cancel()
+    {
+        Fiber to_destroy = std::move(fiber);
+        connection = nullptr;
+    }
+
+    int getFileDescriptor() const { return epoll.getFileDescriptor(); }
+
+private:
+    /// When epoll file descriptor is ready, check if it's an expired timeout.
+    /// Return false if receive timeout expired and socket is not ready, return true otherwise.
+    bool checkReceiveTimeout()
+    {
+        bool is_socket_ready = false;
+        bool is_pipe_ready = false;
+        bool is_receive_timeout_expired = false;
+
+        epoll_event events[2];
+        events[0].data.fd = events[1].data.fd = -1;
+        size_t ready_count = epoll.getManyReady(2, events, true);
+
+        for (size_t i = 0; i != ready_count; ++i)
+        {
+            if (events[i].data.fd == connection->getSocket()->impl()->sockfd())
+                is_socket_ready = true;
+            if (events[i].data.fd == pipe_fd[0])
+                is_pipe_ready = true;
+            if (events[i].data.fd == receive_timeout.getDescriptor())
+                is_receive_timeout_expired = true;
+        }
+
+        if (is_pipe_ready)
+        {
+            drainPipe();
+            return true;
+        }
+
+        if (is_receive_timeout_expired && !is_socket_ready)
+        {
+            receive_timeout.reset();
+            return false;
+        }
+
+        return true;
+    }
+
+    void drainPipe()
+    {
+        uint64_t buf;
+        while (true)
+        {
+            ssize_t res = read(pipe_fd[0], &buf, sizeof(buf));
+            if (res < 0)
+            {
+                if (errno == EAGAIN)
+                    break;
+
+                if (errno != EINTR)
+                    throwFromErrno("Cannot drain pipe_fd", ErrorCodes::CANNOT_READ_FROM_SOCKET);
+            }
+        }
+    }
+
+    struct Routine
+    {
+        PacketReceiver & receiver;
+
+        struct ReadCallback
+        {
+            PacketReceiver & receiver;
+            Fiber & sink;
+
+            void operator()(int, const Poco::Timespan & timeout, const std::string &)
+            {
+                receiver.receive_timeout.setRelative(timeout);
+                receiver.is_read_in_process = true;
+                sink = std::move(sink).resume();
+                receiver.is_read_in_process = false;
+                receiver.receive_timeout.reset();
+            }
+        };
+
+        Fiber operator()(Fiber && sink)
+        {
+            try
+            {
+                while (true)
+                {
+                    {
+                        AsyncCallbackSetter async_setter(receiver.connection, ReadCallback{receiver, sink});
```

Comment:
`AsyncCallbackSetter` is a lightweight structure, cheap to set per packet.
But do we need to reset it every time? Why not to store it as `Routine` field...

Reply:
I just thought that it's not good to remain this `ReadCallback` in connection after received packet, because (theoretically) we can use this connection for something else, but maybe it was bad logic here. What do you think? Should I move it to `Routine` field?

Reply:
Hm, maybe it is true. In case if we store it in `Routine`, this `Routine` can be destroyed later then connection is reused by something else.

Reply:
Let's keep it in callback then